### PR TITLE
support `ANY()` op

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Which issue does this PR close?

Half of #6602 - the overwhelmingly common half.

## Rationale for this change

Trying to support as much postgreSQL syntax as possible, `ANY()` queries are fairly common.

## What changes are included in this PR?

* Support the `ANY()` operator, as in `'needle'=ANY(haystack)`
* Rename the arguments to `array_has` to be less confusing
* tests for `ANY()`
* Some tweaks to test setup to support the above test

## Are these changes tested?

Yes, I can add more if you want.

## Are there any user-facing changes?

Common queries with `ANY()` should now work

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Thank you @samuelcolvin  -- this looks quite good to me 

The only thing I think it needs is a .slt level test

Can you please add one? 

Here are the instructions: https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest

Ideally you should be able to extend one of the existing test files in https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest/test_files 

Maybe somewhere in https://github.com/apache/datafusion/blob/bddb6415a50746d2803dd908d19c3758952d74f9/datafusion/sqllogictest/test_files/array.slt#L4973-L4979 ?

File: datafusion/expr/src/test/function_stub.rs [hunk 1]
```diff
@@ -43,15 +43,13 @@ macro_rules! create_func {
             /// Singleton instance of [$UDAF], ensures the UDAF is only created once
             /// named STATIC_$(UDAF). For example `STATIC_FirstValue`
             #[allow(non_upper_case_globals)]
-            static [< STATIC_ $UDAF >]: std::sync::OnceLock<std::sync::Arc<crate::AggregateUDF>> =
```
Comment (jayzhan211) [inline datafusion/expr/src/test/function_stub.rs:46]: The reason why we have full path in macro is that we can avoid repeatedly importing where ever we have macro

Comment (samuelcolvin) [reply datafusion/expr/src/test/function_stub.rs:46]: I know, but this macro isn't public.

Comment (samuelcolvin) [reply datafusion/expr/src/test/function_stub.rs]: removed anyway.

File: datafusion/sql/src/expr/mod.rs [hunk 2]
```diff
@@ -631,6 +631,31 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                compare_op,
+                right,
+            } => {
+                if compare_op != BinaryOperator::Eq {
+                    return plan_err!(
+                        "Unsupported AnyOp: {compare_op}, only '=' is supported"
+                    );
+                }
+
+                let Some(fun) = self.context_provider.get_function_meta("array_has")
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply datafusion/sql/src/expr/mod.rs]: done.

File: datafusion/functions-nested/src/planner.rs [hunk 3]
```diff
@@ -102,7 +102,7 @@ impl ExprPlanner for NestedFunctionPlanner {
 
     fn plan_make_map(&self, args: Vec<Expr>) -> Result<PlannerResult<Vec<Expr>>> {
         if args.len() % 2 != 0 {
-            return exec_err!("make_map requires an even number of arguments");
+            return plan_err!("make_map requires an even number of arguments");
```
Comment (samuelcolvin) [inline datafusion/functions-nested/src/planner.rs:105]: drive-by change, since this is a "plan" method I assume a "plan error" is more appropriate?

**Review by samuelcolvin — COMMENTED** [review]
I think this is ready.

**Review by alamb — APPROVED** [review]
Thank you @samuelcolvin  -- looks great to me
